### ISSUE TITLE
Update documentation on `dynamic_update_slice_in_dim` 

### ISSUE THREAD

**Issue by davisyoshida** [google/jax#16592]
I was extremely surprised to find that `start_indices` is implicitly set to 0 for dimensions other than `axis`, rather than requiring that all other dimensions be equal or broadcastable to each other:

```python
>>> z = jnp.zeros((3, 5))
>>> u = jnp.ones((2, 2))
>>> jax.lax.dynamic_update_slice_in_dim(z, u, 1, axis=0)
Array([[0., 0., 0., 0., 0.],
       [1., 1., 0., 0., 0.],
       [1., 1., 0., 0., 0.]], dtype=float32)
```

I think it's non-obvious that this is how the `dynamic_update_*_in_dim` methods would work, so noting it in the documentation would be nice.

This led to me really struggling to track down a bug yesterday, which turned out to be in the following code:

```python
updated = jax.lax.dynamic_update_slice_in_dim(operand, update, start_index, axis=2)
```

Unfortunately, I called this code both with arrays of shape `(batch, head, time, hidden)` _and_ `(batch2, batch, head, time, hidden)`. When I was looking for the bug, I thought this couldn't be the problem since I was assuming it would error if it received "incompatible" shapes. Later when I did find it, I assumed it was just broadcasting since the `time` dimension was size 1. (Thanks for the tip on Twitter @mattjj!)

If this _is_ an unintended use of `dynamic_update_slice_in_dim`, it might be nice to do a dimension check, but maybe people are relying on this behavior.

**Comment by jakevdp** [comment]
Hi - thanks for the report, and sorry for the unclear semantics here. Can you say more about what your expected result would be? Were you expecting an error because `u.shape[1] != z.shape[1]`?

**Comment by davisyoshida** [comment]
Yeah I think I was expecting the semantics to be something like numpy slice updates, so for axis=2:

```python
slice = slice(start_index, start_index + update.shape[2])
operand[:, :, slice] = update
```

I think another way this might trip someone up is if they were expecting broadcasting of size-1 dimensions, but instead got the 0-th index updated in those dimensions instead.

**Comment by jakevdp** [comment]
Makes sense, thanks. I believe this was the intended use-case for this function, I might add some shape validation and see if it breaks any code.

**Comment by jakevdp** [comment]
Re-opening and re-naming to track the possibility of adding input validation to these functions.
